### PR TITLE
corrected insert & upsert doc

### PR DESCRIPTION
**Pull request by mhlmhl**
Clarify when insert and upsert results are returned immediately versus via their callback arguments.

### CODE REVIEW

**Comment by sebakerckhof** [comment]
> The ID is passed to the callback method
> untrusted client code, it will be allowed only if passes any	if there is one, otherwise it is returned directly from `insert`.

This is not correct and contradicting the correct behavior further outlined in the docs:

> if you don't provide a callback, then insert blocks until the database acknowledges the write, or throws an exception if something went wrong. **If you do provide a callback, insert still returns the ID immediately**

**Comment by mhlmhl** [comment]
Something got mangled.  What I wrote is:

 

`insert` will generate a unique ID for the object you pass, insert it

in the database, and return the ID. The ID is passed to the callback method

if there is one, otherwise it is returned directly from `insert`.

 

Testing shows that text you cite is wrong.  If a callback is provided, upsert does not return an ID but instead passes it to the callback.

--------------------------------------------------------------------------

Mark H Linehan

**Comment by sebakerckhof** [comment]
The behavior you describe is only for upsert.
A normal insert will always return the id, whether you pass a callback or not.
Repo showing this behavior: https://github.com/sebakerckhof/insertbehavior

**Comment by mhlmhl** [comment]
I built the attached minimal application to test both upsert and insert via calls to Meteor methods.  It has a calling application that invokes ValidatedMethods for all 4 cases of upsert and insert with and without callbacks in the Meteor methods.

 

*       upsert1: upsert using callback in the Meteor method

*       upsert2: upsert without a callback in the Meteor method

*       insert1: insert using callback in the Meteor method

*       insert2: insert without a callback in the Meteor method

 

I put in lots of console.log outputs, prefixed by “**** “ for those that originate from the method caller and “++++ “ for those from the Meteor method.  I logged all returns and callback results to illustrate what is happening.

 

Summarizing the results:

*       upsert1: the document id is passed to the method callback, and the method upsert result is undefined.  The caller does not receive the document id either via the caller’s callback or as the caller’s result.

*       upsert2: the document id is returned by the method upsert. The caller receives the document id both as the caller’s result and via the caller’s callback.

*       insert1: the method callback gets undefined, and the document id is returned by the method insert. The caller does not receive the document id, either via the caller’s callback or as the caller’s result.

*       insert2: the document id is returned by insert in the method. The caller’s callback receives undefined, and the document id is returned via the caller’s result.

 

So you are right that the method insert always returns the document id, but the following statement in the insert documentation seems to be incorrect. From what I can tell, the insert callback never receives the document id.

 

Once the insert completes (or fails), the callback is called with error and result arguments. In an error case, result is undefined. If the insert is successful, error is undefined and result is the new document ID.

 

What is also confusing is what happens when you invoke upsert or insert within Meteor methods.  As you can see above, of the 4 combinations the document id gets returned to the client only for upsert2 and insert2. The  <https://guide.meteor.com/methods.html#basic> guide discussion of basic Method methods describes getting the returned value from a method as follows:

 

If the Method throws an error, you get that in the first argument of the callback. If the Method succeeds, you get the result in the second argument and the first argument err will be undefined. For more information about errors, see the section below about error handling.

 

The tests that I wrote show that in some cases, the result is returned directly to the method call and in some cases, the call-side callback gets an undefined result.

 

I found some articles that say that Meteor methods also can return Promises.  I briefly tried to make that work, but did not spend sufficient time to figure out what is happening for the particular combination of upsert and callback that I used.

 

Summarizing: I spent hours trying to get the document id back from a Meteor method that invokes upsert.  I felt frustrated because the existing documentation is incomplete and misleading. I’m looking to get it corrected.  I spent more hours putting together the attached test to illustrate the problems.

 

--------------------------------------------------------------------------

Mark H Linehan

**Comment by sebakerckhof** [comment]
Can you share that test application? Because the behavior you describe is different from what I see in the minimal app mentioned in my previous comment ( https://github.com/meteor/docs/pull/645#issuecomment-567854896 ).

In any case, the changes this PR proposes are wrong, even if we go by your findings.

**Comment by mhlmhl** [comment]
You are correct that insert returns the inserted document id.  But, contrary to the documentation, that id is never passed to the callback if there is one. This is the inverse of what happens with upsert when there is a callback.

 

It gets even more confusing when you invoke insert and upsert through Meteor methods.  As you can see in my test results, whether the caller of the Meteor method gets the document id at all, and whether it is returned to the caller as a result or via the caller’s callback is quite inconsistent.

 

I took a look at your sample application.  It tests only 1 of the 4 combinations of insert and upsert with and without a callback.  And it doesn’t test them when invoked via a Meteor method. I wrote my test application to explore all these combinations because I had been trying to develop an application that depended on getting the id back and I couldn’t make it work.

 

I sent the test application as a .zip attachment to my last email.  I have attached it again to this email.  Unzip it into a folder, run “npm i simpl-schema” and then run “meteor”.  See the results in the browser debugger and in the server-side console. If it doesn’t come through as a .zip file, I’ll put it on a shared drive so you can access it.

 

I can update the PR if you feel that is necessary.  I found it difficult because of the submodule structure of the docs repo.

 

I think the documentation error for “upsert” is actually only a minor part of what caused me lots of frustration.  The bigger issue is that the Meteor methods documentation is unclear about how and when results are returned from methods.  Not to mention that there may be undocumented support for “Promise” results.  I do not feel comfortable trying to update this part of the documentation, but an update is really needed.  

 

--------------------------------------------------------------------------

Mark H Linehan

**Comment by sebakerckhof** [comment]
The attachments are lost when you reply to a github issue via email. Easiest is to create a github repo or upload the zip on the PR page ( https://github.com/meteor/docs/pull/645 )

Also, the example I posted does test the callback (see https://github.com/sebakerckhof/insertbehavior/blob/master/server/main.js#L6 which contains the callback). When I run this, the callback is called with the id in the insert call and logged to the server console output, e.g.:
```
id from return bCd8ih3jrAafMsQdk
id from callback bCd8ih3jrAafMsQdk
```

Can you try running that repo and confirm this does not happen for you?

Btw, if you want to get the ID back from a meteor method, what's the reason you want to do this via the callback instead of just the return value?

**Comment by mhlmhl** [comment]
[test.zip](https://github.com/meteor/docs/files/3992133/test.zip)
I've attached my test application.

**Comment by sebakerckhof** [comment]
I think there's a misunderstanding here.

If you call a meteor method using `method.call`, the callback will receive the results of whatever is returned by that method. The callbacks you define in the upsert/insert calls are totally unrelated. They are called asynchronously, so after the method will already have returned its results. So what you return in the callbacks of insert/upsert is not passed down as method results. This is nothing meteor specific, this is just how javascript works. So what you're trying to do in `test.upsert1` will not work.

You could, in theory, wrap the insert/upsert call in a promise and resolve the promise in the callback of the insert/upsert and pass down the promise as method result, but this doesn't have any benefit. The mongo calls do not block the event loop whether you pass a callback or not (because the mongo call happens in a coroutine as implemented by https://github.com/laverdet/node-fibers )

Furthermore, the reason the callback `test.insert1` doesn't seem to work, is because the callback signature is wrong. You pass it a callback signature as if it's an upsert ( `(error, { numberInserted, insertedId })`), but this should be ( `(error, insertedId)`) 

So to summarize:
- Insert works as is described in the documentation
- Upsert callback works like described in the documentation
- Upsert does not return results if there's a callback defined, which is not clear in the documentation

**Comment by mhlmhl** [comment]
Seba,

 

I think it’s not a misunderstanding so much as a lack of documentation.  I went back and reviewed the discussion of Meteor methods in both the Meteor Guide and the Meteor API documentation.

*       The Guide documents one way to get the results from a Meteor.call: via the callback method. All the examples show using the callback to get the result. I can find nothing in the Guide that says the results are returned directly from Meteor.call.

*       The API documentation for Meteor.call does show that it returns a result.  I only found that statement after much searching.

*       Neither document says anything about the use of JS Promises with Meteor methods.

 

I do understand that the callback for  insert or upsert is distinct from the callback for Meteor.call.  What I didn’t think about is that the insert/upsert callbacks are necessarily asynchronous and thus values returned from them do not get back to the client. I can find nothing in the documentation that discusses the interaction of collection callbacks and method callbacks.


Thanks for pointing out my error with test.insert1.  But it doesn’t change the points that I am trying to make.  In addition to your summary, I suggest that the Guide should:

 

*       In the “Calling” section, where it discusses receiving the method result via the callback, it should make it clear that the result is also available via the returned value of Meteor.call.  Otherwise, the reasonable assumption is that the ONLY way to get the result is via the callback.

*       Somewhere, the Guide should make it clear that the results received by callbacks of Collection functions that are invoked inside Meteor methods cannot be passed back to code that invokes Meteor.call because of the asynchronous nature of the Collection functions.

 

--------------------------------------------------------------------------

Mark H Linehan

**Comment by sebakerckhof** [comment]
Maybe I didn't explain it clearly, but the result of `Method.call` is only available via the callback, but the result the callback receives is what the method's implementation returns.

What I was trying to point out is that the callback of the upsert call on the collection is unrelated. Not necessarily because of the asynchronous nature, even if the callbacks were called synchronous, it wouldn't make sense for that to be the method call returns value, since you can do multiple in one method call.

Imagine a Meteor method like:

```js
new ValidatedMethod({
  name: "insert2Documents",
  validate: null,
  run() {
     collection.insert(document1, (error, insertedId) => insertedId);
     collection.insert(document2, (error, insertedId) => insertedId);
     return 'foo';
  }
});
```

What would be the return value of this method: The return of the callback of the first insert, the second insert or the return value from the `run` ?
Maybe the guide could have methods that always return a value to make it more clear. Currently most examples return undefined.

**Comment by mhlmhl** [comment]
Updated version of my test application.
[test.zip](https://github.com/meteor/docs/files/3993076/test.zip)

**Comment by mhlmhl** [comment]
It appears to me that this example returns “foo”.

 

This statement is not true:

 

the result of Method.call is only available via the callback, but the result the callback receives is what the method's implementation returns

 

The  <https://docs.meteor.com/api/methods.html#Meteor-call> Meteor.call API documentation says:

 

If you do not pass a callback on the server, the method invocation will block until the method is complete. It will eventually return the return value of the method, or it will throw an exception if the method threw an exception.

 

I made a couple of changes to my test application, and have reposted it to the PR page.  The new version shows that the Method.call returns the result from the Meteor method in all cases, whether the Meteor.call has a callback or not.  So the API documentation statement that I quote above is not quite right. Not if you assume that the conditional at the beginning of the paragraph “If you do not pass a callback …” applies to the entire paragraph rather than just the first sentence.

 

When a Meteor method invokes a collection method with a callback, I can understand that the collection callbacks run asynchronously.  In that case, I can understand why the collection method callback results do not become the return value of the Meteor.call, nor are they passed to the callback of the Meteor.call.  I wish the documentation pointed that out.

 

If the collection method callbacks were synchronous, then if any of them did a “return”, that would give the return value.  But since they are asynchronous, “return” actually returns to the Fiber or Promise that executes the callback. At least when using Promise, the disposition of the return is standardized.  But the Meteor method documentation does not discuss how Promises may be used.

--------------------------------------------------------------------------

Mark H Linehan

**Comment by sebakerckhof** [comment]
Your new application only tests method calls return values on the client-side. But the part of the documentation you quote is about the server-side. The documentation is correct regarding server-side calls.

So to add to the summary:
- Server-side method calls will return if there's no callback. Documentation is correct, but the guide also mentions you typically shouldn't call methods server side.
- Client side the method always returns, but the return value is the value of the stub, and is therefore typically also not safe to use. The documentation also describes this correctly as well (I was wrong here, but again note that using the stub's return value is not safe in the general case).

Regarding promises, there's nothing meteor specific about them. It handles promises just as JS normally works. Same for the callbacks, since callbacks in JS should always be called asynchronously.

**Comment by sebakerckhof** [comment]
So all in all, the documentation is correct on all cases except for the upsert return value when there is a callback. I do agree it is confusing having all these different possibilities of doing the same thing, especially when behavior differs between server & client in subtle ways.
While there are use cases to justify these different possibilities, I think we could emphasize the 99% use case which is:
- Don't use callbacks for the collection methods
- Use callbacks for method calls

**Comment by mhlmhl** [comment]
Your statement that “Your new application only tests method calls return values on the client-side.” is not correct.  The server-side Meteor.method code prints the results of each server side call.  For example:

 

  console.log(

    "**** upsert1 returned result: " +

      testUpsert1.call({ _id: null, name: "name" }, (error, result) => {

        if (error) {

          console.log("**** " + error);

        } else {

          console.log(`**** upsert1 callback result is ${result}`);

        }

      })

  );

 

I do not understand “the guide also mentions you typically shouldn't call methods server side”. The  <https://guide.meteor.com/security.html#allow-deny> Avoid allow/deny section of the Guide says:

 

Given the points above, we recommend that all Meteor apps should use Methods to accept data input from the client, and restrict the arguments accepted by each Method as tightly as possible.

 

… and then goes on to show how to deny all client-side collection changes.  Given that, there is no choice other than using Meteor methods to do collection updates.

 

It is up to Meteor whether it wants to describe the interaction of Meteor calls and Promises. If Meteor wants to leave that up to developers to figure out, that is certainly legitimate.  I think this whole topic is a barrier to usage, one that discourages developers from using Meteor. And one where more guidance, not less, would be very helpful. 

 

--------------------------------------------------------------------------

Mark H Linehan

**Comment by mhlmhl** [comment]
I agree about emphasizing the 99% case.

 

I think the documentation is technically correct but actually very confusing. Something should be done to clarify it.


What would really help are examples that spell out how to use insert and upsert via collection methods. Developers much prefer to cut-and-paste working code than to read through pages and pages of confusing and very subtle text.

 

--------------------------------------------------------------------------

Mark H Linehan

**Comment by sebakerckhof** [comment]
> Your statement that “Your new application only tests method calls return values on the client-side.” is not correct.

Yes it is. What is meant with a server-side call, is that you do a `method.call` from the server code of your application. So for example doing a `method.call` in the `server/main.js` file.
And this is what the guide recommends not to do:
https://guide.meteor.com/methods.html#basic
Under 'calling': 

> This Method is callable from the client and server using Meteor.call. Note that you should only use a Method in the case where some code needs to be callable from the client; if you just want to modularize code that is only going to be called from the server, use a regular JavaScript function, not a Method.

-------
> What would really help are examples that spell out how to use insert and upsert via collection methods

But the examples (e.g. https://guide.meteor.com/methods.html#basic ) in the docs and the guide actually use the 99% case (doing the `.call` with a callback, and the inserts/updates/upserts don't have callbacks). It could maybe be more clear if we return the result of the insert/update/upserts from the method, but this is the first time I've seen confusion about this either on github or in the forums.

So the changes I propose would be:
- Add return values to the method examples
- Clarify that there's no return value for an upsert if you give it a callback.

**Comment by mhlmhl** [comment]
In my test application, the method calls are performed from /client/main.js, which executes on the client side. The server/main.js contains only this:

 

import { Meteor } from "meteor/meteor";

import { testUpdate } from "../imports/api/methods";

 

Meteor.startup(() => {

  // code to run on server at startup

});

 

So I think you are misreading my code.

 

The concern I had with the examples is exactly that they don’t show capturing the return values.

 

I agree with the two changes you propose.  I also wish there were some text that makes it clear that when you use callbacks on collection operations within Meteor methods, you can’t return the results from those operations to client-side code.  Lastly, I wish there were a discussion of using Promises with Meteor methods. 

 

--------------------------------------------------------------------------

Mark H Linehan

**Comment by sebakerckhof** [comment]
> In my test application, the method calls are performed from /client/main.js, which executes on the client side. The server/main.js contains only this:

That's exactly what I mean. For there to be server-side calls (which means calling a method FROM the server side, not the server-side execution of a client-side method call), there should be method calls in, for example, the server/main.js file.

This was a reply to your statement:
> Your statement that “Your new application only tests method calls return values on the client-side.” is not correct.  The server-side Meteor.method code prints the results of each server side call.

We could mention in the guide you can return a promise. I think mentioning you can't get the return value from an upsert callback in a method would just add confusion. Nobody with JS experience would expect this to work this way, since that's just not how callback functions work.

**Comment by mhlmhl** [comment]
Regarding “server-side” if that’s the common Meteor usage, then the terminology needs to be better defined. Meteor methods are defined as executing on both the client and the server, and the server-side execution of Meteor methods certainly sounds server-side to me.

 

Regarding the usage of collection callbacks in Meteor methods.  Perhaps the right thing to do is to caution users not expect that the result of a collection callback in that context can be returned to the client.

 

--------------------------------------------------------------------------

Mark H Linehan

**Comment by sebakerckhof** [comment]
While you're right regarding server side method ececution, but you're confusing it with method call. Method call != Method execution. And if you read back the part of the documentation you quoted ( in this comment: https://github.com/meteor/docs/pull/645#issuecomment-568308647 ) and my answer on that is about server side method calls.

**Comment by filipenevola** [comment]
Hi, as we have a conflict here and also a pending discussion I'm closing this PR.
